### ISSUE TITLE
🖐️ Any tips (or lessons learned) on the best way to integrate Dagger into existing CI

### ISSUE THREAD

**Issue by purpleclay** [dagger/dagger#6920]
### What happened? What did you expect to happen?

I really like the look of Dagger and I am about to take the plunge and migrate my existing GitLab workflows. But I admit, I am a little stumped as to what approach I should take. The documentation focuses heavily on Dagger Modules, so I imagine that is the preferred path.

I am in the process of breaking down my existing GitLab pipeline into composable Dagger modules. And this is where the confusion sets in.

Should I:

1. Keep the same approach to my pipeline within GitLab and literally substitute all jobs with `dagger call` commands. This keeps all the heavy lifting in GitLab, but it means I still operate in a world where any future changes to the pipeline require me to push to GitLab and wait to see if it works as expected (sad times). Trust me, I am not a big fan of GitLab YAML (`rules`, `exists`, `changes` etc).
2. Update my project to include its own Dagger module. Install all of the other Dagger modules as dependencies and then wrap them behind useful functions such as `CI()` or `Release()`. Flags can be provided to modify behaviour when building, say, a `tag` or `merge request`.

Writing this and then reading it back immediately leads me to option 2. However, I would greatly appreciate any feedback from the Dagger team or lessons learned from the Dagger Community.

**Comment by purpleclay** [comment]
I have just found this example project: https://github.com/kpenfound/greetings-api

**Comment by kpenfound** [comment]
Hi @purpleclay , great question. That was originally used before you could pass a git url for a Directory type (very early on). I've just updated the workflows to call the `ci` function: https://github.com/kpenfound/greetings-api/blob/main/.github/workflows/test.yml#L14

**Comment by purpleclay** [comment]
I get to deep dive into Dagger tomorrow for two days, so will attempt option 2 listed above 🤞

**Comment by kpenfound** [comment]
Awesome! Yes option 2 sounds like the way I'd go. Keep functions small and focused on small tasks, and then wrap those in larger functions

**Comment by purpleclay** [comment]
@kpenfound It has been going well, but I have just hit a snag trying to write a Dagger module for one critical part of my pipeline. I use a tool I authored called [nsv](https://github.com/purpleclay/nsv) to tag the repository with the next calculated semantic version. For this to work, it needs access to the project's underlying `.git` directory.

This is blocked by default by Dagger. It doesn't look as if there is a way around this, so I may have to write this as a native GitLab job for the time being. 😢 

I have posted this onto Discord, https://discord.com/channels/707636530424053791/1030538312508776540/threads/1222886821654958110
